### PR TITLE
FLUME-3182 add support for SSL/TLS for syslog (tcp) sources

### PR DESCRIPTION
**Pull request by szaboferee**
Adding SSL/TLS support for syslog and multi port syslog sources
This change also contains some refactoring to avoid code duplication in
sources with SSL/TLS support

As SSL/TLS handling is refactored it is convinient to
add the possibility to specify enabled protocol list FLUME-3275 and
add the possibility to specify enabled cipher suite list FLUME-3276

### CODE REVIEW

**Comment by szaboferee** [comment]
some documentation is in progress, I will update the PR with that and the review fixes later

File: flume-ng-core/src/main/java/org/apache/flume/source/SslContextAwareAbstractSource.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+  protected SSLParameters getSSLParametersWithEnabledProtocolsAndCipherSuites(
+      SSLParameters sslParameters) {
+    return new SSLParameters(
+        getEnabledCipherSuites(sslParameters.getCipherSuites()),
+        getEnabledProtocols(sslParameters.getProtocols())
+    );
+  }
+
+  private String[] getEnabledProtocols(String[] enabledProtocols) {
```
Comment (turcsanyip) [inline flume-ng-core/src/main/java/org/apache/flume/source/SslContextAwareAbstractSource.java]: The method name is a bit misleading because the input parameter is also 'enabled protocols'.
I would rename it to getFilteredProtocols (+ getFilteredCipherSuites).

File: flume-ng-core/src/main/java/org/apache/flume/source/SslContextAwareAbstractSource.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    return () -> getSslEngine(useClientMode);
+  }
+
+  protected SSLParameters getSslParametersForContext(SSLContext sslContext) {
+    SSLParameters sslParameters = sslContext.getDefaultSSLParameters();
+
+    return getSSLParametersWithEnabledProtocolsAndCipherSuites(sslParameters);
+  }
+
+  protected SSLParameters getSSLParametersWithEnabledProtocolsAndCipherSuites(
```
Comment (turcsanyip) [inline flume-ng-core/src/main/java/org/apache/flume/source/SslContextAwareAbstractSource.java]: The method name is a bit complicated. On the other hand, the name of the previous method does not describe exactly what it does (following the same logic that should be getSSLParametersWithEnabledProtocolsAndCipherSuitesForContext).
So it is hard to follow what these methods do.

I would simply implement the overloaded version of getEnabledProtocols / getEnabledCipherSuites with SSLParameters input parameter and extract the filtered lists from that object.

File: flume-ng-sdk/src/main/java/org/apache/flume/api/NettyAvroRpcClient.java [hunk 3]
```diff
@@ -96,7 +99,10 @@
   private String truststore;
   private String truststorePassword;
   private String truststoreType;
-  private final List<String> excludeProtocols = new LinkedList<String>();
+  private final Set<String> excludeProtocols = new LinkedHashSet<>(Arrays.asList("SSLv3"));
```
Comment (turcsanyip) [inline flume-ng-sdk/src/main/java/org/apache/flume/api/NettyAvroRpcClient.java]: I think these fields and also some logic (eg. parseList()) could moved to AbstractRPCClient.

File: flume-ng-core/src/test/java/org/apache/flume/source/TestMultiportSyslogTCPSource.java [hunk 4]
```diff
@@ -155,6 +163,103 @@ private static final int getFreePort() throws IOException {
     return portList;
   }
 
+  private List<Integer> testNPortsSSL(MultiportSyslogTCPSource source, Channel channel,
```
Comment (turcsanyip) [inline flume-ng-core/src/test/java/org/apache/flume/source/TestMultiportSyslogTCPSource.java]: This and testNPorts methods have a lot of common code. As far as I see, testNPortsSSL contains the same code as  testNPorts with some additional logic.
Wouldn't it be possible to consolidate this code duplication?

File: flume-ng-core/src/main/java/org/apache/flume/source/AvroSource.java [hunk 5]
```diff
@@ -455,52 +411,19 @@ private PatternRule generateRule(String patternRuleDefinition) throws FlumeExcep
...
+    public AdvancedChannelPipelineFactory(boolean enableCompression, boolean enableIpFilter,
+        String patternRuleConfigDefinition, Supplier<Optional<SSLEngine>> sslEngineProvider) {
       this.enableCompression = enableCompression;
-      this.enableSsl = enableSsl;
-      this.keystore = keystore;
-      this.keystorePassword = keystorePassword;
-      this.keystoreType = keystoreType;
       this.enableIpFilter = enableIpFilter;
       this.patternRuleConfigDefinition = patternRuleConfigDefinition;
+      this.sslEngineSupplier = sslEngineProvider;
```
Comment (turcsanyip) [inline flume-ng-core/src/main/java/org/apache/flume/source/AvroSource.java]: sslEngineProvider => sslEngineSupplier

**Review by turcsanyip — CHANGES_REQUESTED** [review]
Generally looks good to me. Thanks for your contribution.
Please consider my suggestions regarding method names / restructuring and eliminating duplicated code sections (see my comments inline).

**Comment by majorendre** [comment]
It looks good to me apart from the bit of code duplication @turcsanyip mentioned.

**Review by majorendre — APPROVED** [review]
It looks good to me apart from the bit of code duplication @turcsanyip mentioned.

**Review by turcsanyip — APPROVED** [review]
It looks good to me now.
Thanks for the fixes.

**Comment by majorendre** [comment]
@szaboferee  Thanks for the changes, all looks good.
